plugins/inputs/redis_sentinel/redis_sentinel.go

```diff
@@ -0,0 +1,464 @@
+	if err != nil {
+		return err
+	}
+
+	for i, serv := range r.Servers {
+		u, err := url.Parse(serv)
+		if err != nil {
+			return fmt.Errorf("unable to parse to address %q: %v", serv, err)
+		}
+
+		if u.Scheme != "tcp" && u.Scheme != "unix" {
+			return fmt.Errorf("invalid scheme %q. expected tcp or unix", u.Scheme)
+		}
+
+		address := u.Host
+		if u.Scheme == "unix" {
+			address = u.Path
+		}
+
+		sentinel := redis.NewSentinelClient(
+			&redis.Options{
+				Addr:      address,
+				Password:  r.Password,
+				Network:   u.Scheme,
+				PoolSize:  1,
+				TLSConfig: tlsConfig,
+			},
+		)
+
+		tags := map[string]string{}
+		if u.Scheme == "unix" {
+			tags["socket"] = u.Path
+		} else {
+			tags["source"] = u.Hostname()
+			tags["port"] = u.Port()
+		}
+
+		r.clients[i] = &RedisSentinelClient{
+			sentinel: sentinel,
+			tags:     tags,
+		}
+	}
+
+	return nil
+}
+
+// Redis list format has string key/values adjacent, so convert to a map for easier use
+func toMap(vals []interface{}) map[string]string {
+	m := make(map[string]string)
+
+	for idx := 0; idx < len(vals)-1; idx += 2 {
+		key, keyOk := vals[idx].(string)
+		value, valueOk := vals[idx+1].(string)
+
+		if keyOk && valueOk {
+			m[key] = value
+		}
+	}
+
+	return m
+}
+
+func castFieldValue(value string, fieldType configFieldType) (interface{}, error) {
+	var castedValue interface{}
+	var err error
+
+	switch fieldType {
+	case configFieldTypeFloat:
+		castedValue, err = strconv.ParseFloat(value, 64)
+	case configFieldTypeInteger:
+		castedValue, err = strconv.ParseInt(value, 10, 64)
+	case configFieldTypeString:
+		castedValue = value
+	default:
+		return nil, fmt.Errorf("unsupported field type %v", fieldType)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed cating value %v: %v", value, err)
+	}
+
+	return castedValue, nil
+}
+
+func prepareFieldValues(
+	fields map[string]string,
+	configFieldTypeMap map[string]configFieldType,
+) (map[string]interface{}, error) {
+	preparedFields := make(map[string]interface{})
+
+	for key, val := range fields {
+		key = strings.Replace(key, "-", "_", -1)
+
+		valType, valTypeOk := configFieldTypeMap[key]
+
+		if !valTypeOk {
+			continue
+		}
+
+		castedVal, castedValErr := castFieldValue(val, valType)
+
+		if castedValErr != nil {
+			return nil, castedValErr
+		}
+
+		preparedFields[key] = castedVal
+	}
+
+	return preparedFields, nil
+}
+
+// Reads stats from all configured servers accumulates stats.
+// Returns one of the errors encountered while gather stats (if any).
+func (r *RedisSentinel) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for _, client := range r.clients {
+		wg.Add(1)
+
+		go func(acc telegraf.Accumulator, client *RedisSentinelClient) {
+			defer wg.Done()
+
+			gatherMasterStats(acc, client)
+			gatherInfoStats(acc, client)
+		}(acc, client)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func gatherInfoStats(acc telegraf.Accumulator, client *RedisSentinelClient) {
```

Comment:
How about making this a client member function and maybe even return the errors here and handle them in the level above?
```suggestion
func (client *RedisSentinelClient) gatherInfoStats(acc telegraf.Accumulator) error {
```

This would remove the
```golang
		acc.AddError(infoErr)
		return
```
patterns and you could simply return the error directly. Same for `gatherMasterStats()` et al.

Reply:
@srebhan Addressed in https://github.com/influxdata/telegraf/pull/10042/commits/81760105e877ad7933569287ed47aca43595d8f5. All error recording now happens in the main `Gather` method.